### DISCUSSION TITLE
Incorrect email notifications

### DISCUSSION THREAD

**Discussion by rambant777** [wazuh/wazuh#27245] [Q&A] [answered]
In the web-interface the rule works correctly, when a user has entered the password incorrectly more than 20 times a notification is sent.
But when I add the option 
```  
<options>alert_by_email</options>
```  
All authentication related notifications that are not related to this rule start to arrive
```  
<rule id=“100085” level=“10” frequency=“20” timeframe=“600”>
    <if_matched_sid>60122</if_matched_sid>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
    <description>Failed login 20+ for 10min</description>
   </rule>
```
 How can this situation be corrected?

**Comment by xkazinx** [comment]
Hello @rambant777.

Are you still experiencing this issue? 
You only want the `100085` rule to be sent via email?

**Reply by rambant777** [reply]
> Hello @rambant777.
> 
> Are you still experiencing this issue? You only want the `100085` rule to be sent via email?

Hi @xkazinx!
To be duplicated to email so you don't have to go into wazuh to view logs.
But when I enable the alert_by_email option, events that are not even related to the rule 100085 start to arrive.

**Reply by xkazinx** [reply]
You can use the `wazuh-maild` to send emails notifications based on the alerts that are triggered in the manager.

In the following links you can find more information about setting up the mail daemon:

- https://wazuh.com/blog/how-to-send-email-notifications-with-wazuh/
- https://documentation.wazuh.com/current/user-manual/manager/manual-email-report/index.html

In your case, you will need to use the `rule_id` option to specify which alerts will be sent by email.

```xml
<email_alerts>
  <email_to>you@example.com</email_to>
  <rule_id>5402, 5403</rule_id>
  <do_not_delay />
</email_alerts>
``` 

In the mail daemon behavior, the `minimum severity level`  takes precedence over the configured `rule_id`

In order to avoid this behavior, I followed the next steps:
- I created a custom rule to catch the rule `5402`. This is because the wazuh rules may be overwritten in the update process, so you will need to configure it again, whether user rules are not overwritten. More info here https://documentation.wazuh.com/current/user-manual/ruleset/index.html. This is the rule that I added in `/var/ossec/etc/rules/local_rules.xml`:
```xml
<!-- Local rules -->
<group name="local,syslog,sshd,">
  <rule id="100001" level="5">
    <if_sid>5402</if_sid>
    <description>Successful sudo to ROOT executed.</description>
    <mitre>
      <id>T1548.003</id>
    </mitre>
    <options>alert_by_email</options>
    <group>pci_dss_10.2.5,pci_dss_10.2.2,gpg13_7.6,gpg13_7.8,gpg13_7.13,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_AU.14,nist_800_53_AC.7,nist_800_53_AC.6,tsc_CC6.8,tsc_CC7.2,tsc_CC7.3,</group>
  </rule>
</group>              
```
- Then I used the following ossec.conf
```xml
<ossec_config>
  <global>
    <jsonout_output>yes</jsonout_output>
    <alerts_log>yes</alerts_log>
    <logall>no</logall>
    <logall_json>yes</logall_json>
    <email_notification>yes</email_notification>
    <smtp_server>localhost</smtp_server>
    <email_from>wazuh@example.wazuh.com</email_from>
    <email_to>recipient@example.wazuh.com</email_to>
    <email_maxperhour>12</email_maxperhour>
    <email_log_source>alerts.log</email_log_source>
    <agents_disconnection_time>10m</agents_disconnection_time>
    <agents_disconnection_alert_time>0</agents_disconnection_alert_time>
  </global>

  <alerts>
    <log_alert_level>3</log_alert_level>
    <email_alert_level>12</email_alert_level>
  </alerts>


  <!-- Choose between "plain", "json", or "plain,json" for the format of internal logs -->
  <logging>
    <log_format>plain</log_format>
  </logging>
```
- Finally, to check the behavior, I run the mail binary in debug mode:
```
2022/03/16 10:04:15 wazuh-maild[5601] maild.c:258 at OS_Run(): INFO: Getting alerts in log format.
2022/03/16 10:04:25 wazuh-maild[5601] os_maild_client.c:218 at OS_RecvMailQ(): DEBUG: OS_RecvMailQ: mail->body[
Wazuh Notification.
2022 Mar 16 10:04:24

Received From: ubuntumanager->/var/log/auth.log
Rule: 100001 fired (level 5) -> "Successful sudo to ROOT executed."
User: root
Portion of the log(s):

Mar 16 10:04:22 ubuntumanager sudo:  vagrant : TTY=pts/2 ; PWD=/home/vagrant ; USER=root ; COMMAND=/usr/bin/su
tty: pts/2
pwd: /home/vagrant
command: /usr/bin/su
 --END OF NOTIFICATION
]
```
To run the mail daemon in foreground and in debug mode, follow the next steps:
- Once you have configured the wazuh manager, restart it as normal.
- After that, you need to kill the maild process, you can use the following command: `pkill wazuh-maild`.
- Then, run the mail daemon in the foreground and debug mode: `/var/ossec/bin/wazuh-maild -fdd`. You can find the meaning of the options here https://documentation.wazuh.com/current/user-manual/reference/daemons/wazuh-maild.html

Hope you find this useful

Source of this information: https://github.com/wazuh/wazuh/issues/12717
See more: https://github.com/wazuh/wazuh/issues?q=%22send%20email%22

**Reply by rambant777** [reply]
Hi @xkazinx !
I get alerts under a different set of rules. The problem is that this is the rule that alerts come with normal authorization, not when there have been 20+ failed attempts.
So I'm thinking maybe I didn't specify some necessary line in the code of the custom rule?

**Reply by xkazinx** [reply]
Can you share your `/var/ossec/etc/ossec.conf` and `/var/ossec/etc/rules/local_rules.xml` file contents?

**Reply by rambant777** [reply]
> Can you share your `/var/ossec/etc/ossec.conf` and `/var/ossec/etc/rules/local_rules.xml` file contents?

Hi!
I can't upload the ossec.conf file due to server limitations, my main question is - is the rule built correctly? Because other rules work fine, and notifications to email come without false positives.

**Reply by xkazinx** [reply]
Are you adding the option inside of the `<rule>` scope?
```
<options>alert_by_email</options>
```

Since, if a rule with the `alert_by_email` option is matched inside the `<rule>` scope, it should always generate an email alert, regardless of the minimum level set by `email_alert_level`, which determines which emails to send based on the alert level; but if `alert_by_email` is added in a different scope, it will send alerts from other rules as well. 

If this is not the case, I will go ahead, investigate this further by manually reproducing it, and sharing the logs and steps to configure it.

**Reply by rambant777** [reply]
> Are you adding the option inside of the `<rule>` scope?

Yes, rule:
```
<rule id=“100085” level=“10” frequency=“20” timeframe=“600”>
    <if_matched_sid>60122</if_matched_sid>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
    <options>alert_by_email</options>
    <description>Failed login 20+ for 10min</description>
   </rule>
```
Judging by the rule, events should come when rule 60122 is triggered more than 20 times in 10 minutes.
But there are also normal events related to rule 60122, not events that occurred more than 20 times.

**Reply by xkazinx** [reply]
Please, try the following to determine that there is no conflict between the use of `frequency` and `alert_by_email`:

```
<rule id="100085" level="10" frequency="20" timeframe="600">
    <if_matched_sid>60122</if_matched_sid>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
    <description>Failed login 20+ for 10min</description>
</rule>

<rule id="100086" level="10">
    <if_sid>100085</if_sid>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
    <options>alert_by_email</options>
    <description>Send email for Failed login 20+ for 10min</description>
</rule>
```

where `alert_by_email` is removed from the `100085` rule and placed in the `100086` rule instead, where it's still expected to be triggered after 20 attempts.

**Reply by rambant777** [reply]
> Please, try the following to determine that there is no conflict between the use of `frequency` and `alert_by_email`:

Hi!
I created a child rule as you wrote above, about 30 minutes after restarting wazuh manager started receiving notifications that are not related to the rule 60122 “Account failed to log in."
There are no such events in the alerts of rules 100085 and 100086.

**Reply by xkazinx** [reply]
Okay, just in case, make sure you have removed the `alert_by_email` from the rule `100085`.
I will go ahead and try this on an environment. I will get back to you soon.

**Reply by rambant777** [reply]
> Okay, just in case, make sure you have removed the `alert_by_email` from the rule `100085`.
> I will go ahead and try this on an environment. I will get back to you soon.

Thanks!

**Reply by xkazinx** [reply, marked as answer]
After testing the configuration on my environment, I'm successfully receiving emails when meeting the `frequency` property of `3` matches, for failed `apache2` logins.

`/var/ossec/etc/local_rules.xml`:
```
<rule id="30309" level="9">
    <if_sid>30301</if_sid>
    <id>AH01618|AH01808|AH01790</id>
    <description>Apache: Attempt to login using a non-existent user.</description>
    <group>invalid_login,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gpg13_7.1,gdpr_IV_35.7.d,gdpr_IV_32.2,</group>
</rule>

<rule id="100002" level="9" frequency="3" timeframe="120">
  <if_matched_sid>30309</if_matched_sid>
  <description>Multiple user authentication failures</description>
  <options>alert_by_email</options>
  <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,gpg13_7.1,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_AU.3.1,nist_800_53_IA.10,</group>
</rule>
```

`/var/ossec/etc/ossec.conf`:
```
<ossec_config>
  <global>
    <jsonout_output>yes</jsonout_output>
    <alerts_log>yes</alerts_log>
    <logall>no</logall>
    <logall_json>no</logall_json>
    <email_notification>yes</email_notification>
    <smtp_server>localhost</smtp_server>
    <email_from>xkazzinx@gmail.com</email_from>
    <email_to>xkazzinx@gmail.com</email_to>
    <email_maxperhour>999</email_maxperhour>
    <email_log_source>alerts.log</email_log_source>
    <agents_disconnection_time>10m</agents_disconnection_time>
    <agents_disconnection_alert_time>0</agents_disconnection_alert_time>
    <update_check>yes</update_check>
  </global>

  <alerts>
    <log_alert_level>3</log_alert_level>
    <email_alert_level>10</email_alert_level>
  </alerts>
```

Please, check if there are other custom rules that could be producing your output, as well as comparing the rules I sent with the related ones you have.

**Reply by rambant777** [reply]
> <rule id="30309" level="9">
>     <if_sid>30301</if_sid>
>     <id>AH01618|AH01808|AH01790</id>
>     <description>Apache: Attempt to login using a non-existent user.</description>
>     <group>invalid_login,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gpg13_7.1,gdpr_IV_35.7.d,gdpr_IV_32.2,</group>
> </rule>

Hi!
In my dashboard, I can't see the events related to rules 30301 and 30309. Shouldn't a custom rule be built on multiple alerts from rule 60122?
Why 30301/30309?

**Comment by zoooowy** [comment]
thx
